sycl/doc/extensions/KernelProperties/KernelProperties.asciidoc

```diff
@@ -0,0 +1,519 @@
+  that attributes should not be employed to convey information that cannot be
+  ignored. Many of the attributes in SYCL 2020 convey information that cannot
+  be ignored (e.g. a kernel may only execute correctly with a specific
+  sub-group size, or on devices which have specific aspects).
+
+- Library-only implementations of SYCL cannot reason about attributes at all.
+
+- SMCP implementations of SYCL must ensure that the host and device compiler
+  both understand and agree upon the meaning of each attribute.
+
+- It is complicated (although not impossible) to generate multiple variants of
+  a kernel with different combinations of attributes.
+
+- There is no way to inspect the properties associated with a kernel at
+  compile-time (e.g. via type traits or similar); any such inspection must be
+  performed at run-time and *after* compiling the kernel.
+
+This extension proposes a replacement for these kernel attributes, in the form
+of a `property_list` accepting properties with compile-time constant
+values, to address several of these issues.
+
+== Kernel Properties
+
+The kernel properties below correspond to kernel attributes defined in
+Section 5.8.1 of the SYCL 2020 specification.  Note that deprecated attributes
+(such as `vec_type_hint`) are not included.  The `requires` attribute has been
+renamed `device_has`, to avoid future conflicts with the {cpp}20 keyword.
+
+```c++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+// Corresponds to reqd_work_group_size
+struct work_group_size {
+  template <size_t... Dims>
+  using value_t = property_value<work_group_size, std::integral_constant<size_t, Dims>...>;
+}; // work_group_size
+
+// Corresponds to work_group_size_hint
+struct work_group_size_hint {
+  template <size_t... Dims>
+  using value_t = property_value<work_group_size_hint, std::integral_constant<size_t, Dims>...>;
+}; // work_group_size_hint
+
+// Corresponds to reqd_sub_group_size
+struct sub_group_size {
+  template <uint32_t Size>
+  using value_t = property_value<sub_group_size, std::integral_constant<uint32_t, Size>>;
+}; // sub_group_size
+
+// Corresponds to requires
+struct device_has {
+  template <sycl::aspect... Aspects>
+  using value_t = property_value<device_has, std::integral_constant<sycl::aspect, Aspects>...>;
+}; // device_has
+
+template <size_t... Dims>
+struct property_value<work_group_size, std::integral_constant<size_t, Dims>...> {
+  constexpr size_t operator[](int dim);
+};
+
+template <size_t... Dims>
+struct property_value<work_group_size_hint, std::integral_constant<size_t, Dims>...> {
+  constexpr size_t operator[](int dim);
+};
+
+template <sycl::aspect... Aspects>
+struct property_value<device_has, std::integral_constant<sycl::aspect, Aspects>...> {
+  static constexpr std::array<sycl::aspect, sizeof...(Aspects)> value;
+};
+
+template <size_t... Dims>
+inline constexpr work_group_size::value_t<Dims...> work_group_size_v;
```

Comment:
```suggestion
template <size_t... Dims>
inline constexpr __unspecified__ work_group_size;
```
with `__unspecified__` secretly the `work_group_size_t` above

Reply:
> Quite interesting.
> I propose some shorter implementation by hiding the concrete property types.

Thanks, @keryell.  I didn't add any of the motivation for the naming, here (because I thought it would be covered by the property list extension) but I don't think we can make things completely unspecified or hide too many implementation details.

The user needs to be able to:

- Pass a value with a specific compile-time-constant size when constructing a `property_list`
- Pass a type without a specific compile-time-constant size when querying a `property_list`

For example: `{ sycl::ext::oneapi::sub_group_size_v<8> }`  specifies a sub-group size of 8, and `has_property<sycl::ext::oneapi::sub_group_size>()` returns an object representing the sub-group size.

What you're proposing would require `has_property<sycl::ext::oneapi::sub_group_size_t>()`, which might be less obvious?  Adding @rolandschulz to comment more -- there has been lots of back and forth on potential naming schemes for these two concepts, and he could do a better job of summarizing the options we considered.

One benefit to them not being unspecified is that we can specify that different properties will have different interfaces.  For example, `device_has` models an array of aspects, whereas `work_group_size` is intended to behave more like a `sycl::range`.  (We could still do that and keep most of the class unspecified, just making a case for things not being _completely_ unspecified.)

Reply:
Everything John wrote is correct and covers the main points. I'll just try to structure this into separate decisions we made.

1) A property might have a compile-time value, a runtime-value, or no-value. For a runtime-value the property is a class (`foo`) and the property value is an object of that class (`foo{5}`). A compile-time value has to be encoded as a type templated on the value. There are basically two ways to do that. Either a property and a property-value can be two different types or the property can be a class template and the property-value the instantiated class of that same template. We went with the first option, because
a) it is inconsistent if a compile-time property is a class template but other properties are classes. This way all properties are classes.
b) the syntax of dealing with template template arguments is confusing to many and for those users it is easier if a property is a type.
2) With the choice to have two different types it seemed important to have a clear connection between them. This is why we have `struct foo { template<...> using value_t = property_value<foo, ...>; }`. This allows
a) given some property `T` to refer to the property value as `T::value_t`
b) know that the property-value type of property `T` is of type `property_value<foo, ...>`
It is true that b) isn't necessary. The user could be only allowed to use `T::value_t` in which case it's type could be implementation defined. But we don't think it actually would allow any additional implementation freedom. CRTP isn't useful here. And it would prevent advanced users to write templated expressions involving different property values (`property_value<P, Args...>`). Therefore we only see downsides of not specifying the type.
3) Given that `foo::value_t<3>()` is a bit verbose we added she shortcut `foo_v<3>` for that.

Reply:
Thanks for the clarification.
Now I can see the value (sic) of `value_t` and the type to introspect with `has_property`.
So now we have `foo` and `foo_v`.
But I have the feeling a normal user would use `foo_v` quite more than the type itself.
So instead we could have `foo_t` and `foo`?
But thinking again, if `has_property` can accept also template template, we could have a more value-oriented API with just `foo` as the type and `foo { a, b, c }` as a constructor or deduction guide, where `a`, `b`, `c` are values. The drawback is that it is cumbersome to pass `constexpr` integer for example because they have to be boxed into some `constexpr` type wrapper. Perhaps a literal suffix could help.
At the end your proposal looks like a good tradeoff.
CRTP was just a suggestion of an implementation to avoid typing
```c++
 template <size_t... Dims>
  using value_t = property_value<work_group_size, std::integral_constant<size_t, Dims>...>;
```
, nothing a user should look at.

Reply:
> But I have the feeling a normal user would use foo_v quite more than the type itself.

Yes. `foo` is only needed when querying (`has_property`/`get_property`) whereas `foo_v` is the recommended way to setting properties. And querying is going to be much less common.

> So instead we could have foo_t and foo?

We could reverse it. But I don't think `foo_t` would be good naming. Are you suggesting that `_t` stands for "type"? `foo_t` wouldn't be the same as `decltype(foo<n>)`. Currently the naming makes sense to me. One is the property and the other is the property value (but it isn't an object of the property type). I can't think of good naming if we were to reverse it. And I don't think it is problematic that `foo_v` is going to be used much more than `foo`.

> if has_property can accept also template template, we could have a more value-oriented API

That would solve a key problem I forgot to mention of template template arguments. It isn't possible to have a universal template template argument yet (https://github.com/cplusplus/papers/issues/735). Therefore all property value template arguments would need to be either type or non-type. We expect almost all to be non-type but we don't want to restrict it to non-type. Passing `integral_constant` for non-types would solve this. UDL is the only convenient way of doing that. But this is problematic for a few reasons:
- Constexpr UDL requires to use the `char...` template and there is no constexpr builtin for `atoi/strtoi/from_char` and therefore you end up parsing of the string in the header. Not great for compilation speed.
- UDL requires `using namespace` and you might end up with conflicts. 
- Those UDLs don't seem to be a popular choice for future C++ (https://groups.google.com/a/isocpp.org/g/std-proposals/c/LRDS2e3LhZw/m/GWLCUkltEAAJ?pli=1 https://www.reddit.com/r/cpp/comments/ju3poj/literal_suffix_for_stdintegral_constant/)